libcrux-psq/src/ecdh_binder.rs

```diff
@@ -0,0 +1,310 @@
+//! # ECDH Binder
+//!
+//! This module implements a protocol binder that binds a post-quantum
+//! pre-shared key component to an outer ECDH based key agreement.
+
+use libcrux::aead::{decrypt_detached, Algorithm};
+use libcrux_ecdh::{X25519PrivateKey, X25519PublicKey};
+use std::time::{Duration, SystemTime};
+
+use rand::{CryptoRng, Rng};
+
+use crate::{
+    psq::{PrivateKey, PublicKey},
+    Error, Psk,
+};
+
+const DH_PSK_CONTEXT: &[u8] = b"DH-PSK";
+const DH_PSK_LENGTH: usize = 32;
+
+const AEAD_RECEIVER: &[u8] = b"AEAD-Responder";
+const AEAD_INITIATOR: &[u8] = b"AEAD-Initiator";
+const AEAD_KEY_NONCE: usize = Algorithm::key_size(Algorithm::Chacha20Poly1305)
+    + Algorithm::nonce_size(Algorithm::Chacha20Poly1305);
+
+const AEAD_KEY_LENGTH: usize = Algorithm::key_size(Algorithm::Chacha20Poly1305);
+
+/// An ECDH-bound PSQ encapsulation.
+pub struct ECDHPsk {
+    encapsulation: crate::psq::Ciphertext,
+    initiator_dh_pk: Vec<u8>,
+    aead_mac: (libcrux::aead::Tag, Vec<u8>),
+    psk_ttl: Duration,
+    ts: Duration,
+}
+
+/// Generates a post-quantum pre-shared key (PQ-PSK) bound to an ECDH key.
+///
+/// The encapsulated PSK is valid for the given duration
+/// `psk_ttl`, based on milliseconds since the UNIX epoch until
+/// current system time. Parameter `sctx` is used to
+/// cryptographically bind the generated PSK to a given outer
+/// protocol context and may be considered public.
+pub fn send_ecdh_bound_psq(
+    psq_pk: &PublicKey,
+    receiver_dh_pk: &X25519PublicKey,
+    initiator_dh_sk: &X25519PrivateKey,
+    psk_ttl: Duration,
+    sctx: &[u8],
+    rng: &mut (impl CryptoRng + Rng),
+) -> Result<(Psk, ECDHPsk), Error> {
+    let now = SystemTime::now();
+    let ts = now
+        .duration_since(SystemTime::UNIX_EPOCH)
+        .expect("now cannot be before UNIX_EPOCH");
+    let ts_seconds = ts.as_secs();
+    let ts_subsec_millis = ts.subsec_millis();
+    let mut ts_ttl = ts_seconds.to_be_bytes().to_vec();
+    ts_ttl.extend_from_slice(&ts_subsec_millis.to_be_bytes());
+    ts_ttl.extend_from_slice(&psk_ttl.as_millis().to_be_bytes());
+
+    let (ss_q, encapsulation) = psq_pk.gen_pq_psk(sctx, rng)?;
+    let ss_dh = libcrux_ecdh::x25519_derive(receiver_dh_pk, initiator_dh_sk)
+        .map_err(|_| Error::CryptoError)?;
+
+    // ikm = ss_q || ss_dh
+    let mut ikm = Vec::from(ss_q);
+    ikm.extend_from_slice(&ss_dh.0);
+
+    let prk = libcrux_hkdf::extract(libcrux_hkdf::Algorithm::Sha256, b"", ikm);
```

Comment:
Can you file an issue to add a fixed sized APIs for HKDF, or something where we pass in the output?